### DISCUSSION TITLE
Open midi library with button

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#687] [Q&A]
How do I open the midi library or any other tab in the menubar with an uiButton?
Thank you

**Comment by dobo365** [comment]
What do you mean by Midi Library? The Midi settings (= Midi menu content)?

**Reply by jojogib** [reply]
![midil](https://github.com/user-attachments/assets/cb15f093-3449-4d67-a877-df7cfe50adcc)

**Comment by dobo365** [comment]
Amazing... 
I have been using Ctrlr since long and I'm quite an expert of it, still using it almost every day and never seen this. Hahaha!

I guess this is some kind of catalogue/library of different things.
Don't know how this is populated.
Can only say that Snapshots are abandoned.
Maybe that some other Ctrlr fellow will know more.

Now, we can add an acces to it via code in a newer version of CtrlrX. We have Midi Settings and added Midi Monitor.
Yes, you create a uiButton and then use the property "Internal function this component should execute"
To have a nicer button, you create that nice button then, by Lua, set the value of the uiButton that can be hidden.

**Reply by damiensellier** [reply]
Hi, it's been a while, I'm very busy currently.

It seems that this "MIDI Library" module is gone in the 5.6.x versions. 
I don't even know if it was fully working in 5.3.x 

Since this module is gone, I'll have hard time to connect a uiButton to "internal" function in a future update like we did for the "MIDI monitor" or "MIDI Device Selector". Or I would need to resurect it.

You had it working already @jojogib ?

Damien

This is 5.3.198

![Capture d’écran, le 2024-11-02 à 16 03 43](https://github.com/user-attachments/assets/982e716f-24b7-4ed0-aaeb-9211ea6b52d1)
![Capture d’écran, le 2024-11-02 à 16 03 28](https://github.com/user-attachments/assets/d7ac7690-c109-4555-9ba8-53f9591f0bda)

**Reply by jojogib** [reply]
Im using an older version not 5.6. So there is no easy way to call any menubar feature like (snapshot, midimonitor, midi setting... etc ) in LUA? thank you

**Reply by damiensellier** [reply]
No unfortunately.

**Reply by damiensellier** [reply]
you can call "MIDI Device Selection" from a uiButton though : 

![Capture d’écran, le 2024-11-02 à 19 54 24](https://github.com/user-attachments/assets/681983b0-8b99-4576-bff5-b8365a198fba)

From LUA I don't remember. John @dnaldoog or Dominique @dobo365  can probably tell you.

**Reply by dnaldoog** [reply]
Like @dobo365 I have never even noticed that menu item "MIDI Library" &#x2026; strange - I do know about the _MIDI Device Selection_ option you pointed out @damiensellier but don't know how to access other functions. You can alter all the MIDI options in lua though.

**Comment by dobo365** [comment]
As I said above, in CtrlrX we have Midi Settings and added Midi Monitor.
In 5.3.201, you have only Midi Settings.
Yes, you create a **uiButton** and then use the property "_Internal function this component should execute_" as indicated in the picture posted by Damien. Call that button _MidiSettingsHidden_ and make it Not visible.
To have a nicer button, you create a nice **uiImageButton** button then, by Lua, set the value of the uiButton that can be hidden.

I declare the hidden button in my AssignModulators method
`	btnMidiSettingsHidden = panel:getModulatorByName("MidiSettingsHidden")`

Then I use this OnChange method for the visible button:
```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
MidiSettings_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action while the panel is in bootstrap or program states
	if not isPanelReady() or source==6 or source==1 then
		return
	end	

	-- Display Midi settings popup by triggering the hidden uiButton
	btnMidiSettingsHidden:getComponent():click()

end
```

**Reply by jojogib** [reply]
Yeah i noticed that but in ctrlrx there is no midi library

**Reply by dobo365** [reply]
Well, I'm afraid you will have to make a choice...
Go with 5.3.201 and use your Midi Library but being aware that almost nobody can support you with it and knowing that there won't be any new versions of Ctrlr (reason Damien made CtrlrX so we have at least 2 admins if one of them stops for any reason).
Also knowing that the creator of Ctrlr removed it when he did his last version 5.6 (Snapshots have also been removed).
Go with 5.3.201 and do your File handling via Lua using the Juce File class (this is what I'm doing)
Go with CtrlrX that we try to improve based on our understanding and time but without Midi Library.
